### PR TITLE
[Representation] One-way bind from mct-include, mct-representation

### PR DESCRIPTION
**Pull request by VWoeltjen**
Utilize a one-way binding strategy for `mct-include` and `mct-representation`, such that expressions in the parent scope of the directive are evaluated and values are passed into the child scope, but no such values are passed from the child back to the parent.

This avoids extra watches introduced by `scope: { key: "=" }` style two-way binding. In particular, the two-way binding approach is incompatible with one-time binding (`::`-prefixed expressions); these are non-assignable, but a watch is still kept within the child scope. Child-to-parent binding is not currently utilized (and is generally hard-to-utilize due to scope inheritance) so these watches are wasted; instead, child-to-parent communication, when needed, has been achieved with properties on passed-in objects, and this is still supported.

Addresses #314
### Author Checklist
1. Changes address original issue? Y &ast;
2. Unit tests included and/or updated with changes? Y
3. Command line build passes? Y
4. Expect to pass code review? Y

&ast; Wording of #314 assumes that #313 will have been addressed first, but prefer to resolve these in the opposite order; should be able to address #313 trivially with these changes.

### CODE REVIEW

**Comment by akhenry** [comment]
I think this is a reasonable change so long as the performance improvements are significant. As you say, the two way binding of non-changing attributes creates unnecessary watches, and given the high usage of mct-representation, even incremental improvements in performance can cumulatively be significant.

I'd like to get an idea of the performance improvement that you saw for this particular change though. It's a departure from the standard Angular way of doing things, and that may cause confusion to developers. It's also additional code to maintain that could have some subtle behavior.

**Comment by VWoeltjen** [comment]
> I think this is a reasonable change so long as the performance improvements are significant. As you say, the two way binding of non-changing attributes creates unnecessary watches, and given the high usage of mct-representation, even incremental improvements in performance can cumulatively be significant.

That's a good question but a tough one to give a clear answer to. Typically we have very little idle time in the application; profiling shows that 80-90% of time is spent in the digest cycle, most of which is spent evaluating compiled Angular expressions.

Since we can't do much to speed up the evaluations themselves, we're left trying to decrease either the number of expressions which get evaluated (watch count), or their frequency (digests / sec). The Angular team's guidance has been to keep watch counts under 2000, and we're way over that in many cases (esp. with large trees) so that's the metric I've been focusing on improving. 

This change alone gives about a 20% reduction in watch count. (My test case is with a taxonomy of around ~200 domain objects visible in the tree; I have 5394 watches without these changes, and 4345 with them.) But the real benefit (demonstrated in #303) comes when you start using [one-time bindings](https://docs.angularjs.org/guide/expression#one-time-binding) for the attributes passed into `mct-representation` - in conjunction with other changes, this allows the watch count associated with the tree to be reduced to zero. The same strategies can also be reused in other places where there's a high watch density.

**Comment by VWoeltjen** [comment]
...the above got me to take a second look at digest _frequency_, which was high. Turns out a misuse of `$interval` was causing ~140 digests to be triggered each second; correcting this usage brings the count down to 6. Will submit a pull request there; in conjunction with watch count reduction this should see some immediate performance improvement.

**Comment by VWoeltjen** [comment]
#370 reduces digests/sec rather dramatically. That should have a much more immediate impact on performance while the application is idle, but watch count reduction still needs to be reduced for cases where digest frequency goes up for legitimate reasons (ongoing user interaction, higher-rate telemetry data, etc)

**Comment by VWoeltjen** [comment]
Some concrete measurements.

Variant | Digests/sec | Watches | Idle (ms) | $digest (ms) | Idle / $digest
----|----|----|----|----|----
master | 96 | 5426 | 54.9 | 9165.5 | 166.95
master + #370 | 11 | 5426 | 10490.8 | 989 | 0.094273
master + #370 + #365 | 12 | 4375 | 14045.9 | 1218.4 | 0.086744

Test case is WARP with a single autoflow tabular view being displayed (not a particularly stressing case.) The immediate benefits of #370 are dramatic; immediate benefits of this change set are more modest. 

We could consider ending the sprint with just #370 if we consider other change sets to be higher-risk changes.

File: platform/representation/src/OneWayBinder.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+         * expression when changes occur.
+         * @param {string} attr the name of the attribute to watch
+         * @param {Function} callback the callback to invoke with new values
+         */
+        OneWayBinder.prototype.watch = function (attr, callback) {
+            var expr = this.attrs[attr];
+            this.unwatches.push(this.parent.$watch(
+                expr,
+                callback,
+                expr && expr[0] === '{'
```
Comment (akhenry) [inline platform/representation/src/OneWayBinder.js]: Perhaps a comment would be good here to explain the purpose of this check? It seems as though the result will be to do a shallow compare on object references and primitive types, and a deep compare on object literals (eg. when 'parameters' is specified). Is the value of an attribute on an object literal ever going to change though?

**Comment by akhenry** [comment]
A lot of thought has gone into this, nice work. Thanks for the profiling, it's really useful.

That said, I'm still a bit ambivalent. It's a marginal performance gain (though reducing the watch count as per Angular's recommendation is a good thing), but it's a bunch of additional code which adds risk and maintenance cost. Moreover, it adds more obfuscation to MctRepresentation which is already frankly bewildering to new developers.

On balance I think we should merge it. It's well thought out, clean code, and a 20% drop in watch count is a good thing.  Just a minor point of clarification above.

**Comment by VWoeltjen** [comment]
I agree with all those points. Due to the ubiquity of `mct-representation` the scope of any risk is fairly large.

Since #370 gives a more dramatic performance improvement with a smaller changeset, maybe we should just go with that for the current release, and defer this change? (@larkin - your thoughts?)

**Comment by VWoeltjen** [comment]
Reassigning to self to address points from code review

**Comment by larkin** [comment]
Hrm, I'd like to review this a bit more thoroughly and see if there's a slightly more angular way to do it.  Based on your numbers, #370 is a huge fix on it's own and this could be pushed out of this sprint and addressed in future sprints.

**Comment by VWoeltjen** [comment]
Moving to next sprint based on above comments

**Comment by larkin** [comment]
Closing as "not going to implement"

**Pull request closed without merging**